ow_power_system/src/PowerSystemPack.cpp

```diff
@@ -0,0 +1,375 @@
+  PredictionHandler handler_4(bus[4], node_names[4]);
+  PredictionHandler handler_5(bus[5], node_names[5]);
+  PredictionHandler handler_6(bus[6], node_names[6]);
+  PredictionHandler handler_7(bus[7], node_names[7]);
+
+  // Get the asynchronous prognoser configuration and create a builder with it.
+  auto config_path = ros::package::getPath("ow_power_system") + "/config/async_prognoser.cfg";
+  ConfigMap config(config_path);
+
+  ModelBasedAsyncPrognoserBuilder builder(std::move(config));
+  builder.setModelName("Battery");
+  builder.setObserverName("UKF");
+  builder.setPredictorName("MC");
+  builder.setLoadEstimatorName("Const");
+
+  // Create the prognosers using the builder that will send predictions using
+  // their corresponding message bus.
+  PCOE::AsyncPrognoser prognoser_0 = builder.build(bus[0], node_names[0], "trajectory");
+  PCOE::AsyncPrognoser prognoser_1 = builder.build(bus[1], node_names[1], "trajectory");
+  PCOE::AsyncPrognoser prognoser_2 = builder.build(bus[2], node_names[2], "trajectory");
+  PCOE::AsyncPrognoser prognoser_3 = builder.build(bus[3], node_names[3], "trajectory");
+  PCOE::AsyncPrognoser prognoser_4 = builder.build(bus[4], node_names[4], "trajectory");
+  PCOE::AsyncPrognoser prognoser_5 = builder.build(bus[5], node_names[5], "trajectory");
+  PCOE::AsyncPrognoser prognoser_6 = builder.build(bus[6], node_names[6], "trajectory");
+  PCOE::AsyncPrognoser prognoser_7 = builder.build(bus[7], node_names[7], "trajectory");
+
+  ROS_INFO_STREAM("Power system pack running.");
+
+  ros::Rate rate(m_gsap_rate_hz);
+
+  // Loop through the PowerSystemNodes to update their values and send them to the bus
+  // to get predictions.
+  while(ros::ok())
+  {
+    ros::spinOnce();
+    for (int i = 0; i < NUM_NODES; i++)
+    {
+      //ROS_INFO_STREAM("Calling RunOnce on node " << i << "..."); TEST
+      nodes[i].RunOnce();
+      nodes[i].GetPowerStats(models[i]);
+
+      // /* DEBUG PRINT
+      if (!(models[i][0] <= 0))
+      {
+        ROS_INFO_STREAM("Node " << i << "  time: " << models[i][0]
+                        << ". power: " << models[i][1] << ".  volts: "
+                        << models[i][2] << ".  temp: " << models[i][3]);
+      }
+      // */
+
+      auto timestamp = START_TIME + std::chrono::milliseconds(static_cast<unsigned>(models[i][0] * 1000));
+
+      // Compile a vector<shared_ptr<DoubleMessage>> and then individually publish
+      // each individual component.
+      std::vector<std::shared_ptr<DoubleMessage>> data_to_publish;
+
+      data_to_publish.push_back(
+        std::make_shared<DoubleMessage>(MessageId::Watts, node_names[i], timestamp, models[i][1]));
+      data_to_publish.push_back(
+        std::make_shared<DoubleMessage>(MessageId::Centigrade, node_names[i], timestamp, models[i][3]));
+      data_to_publish.push_back(
+        std::make_shared<DoubleMessage>(MessageId::Volts, node_names[i], timestamp, models[i][2]));
+
+      std::this_thread::sleep_until(timestamp);
```

Comment:
What is the purpose of this?

Reply:
Assuming you're talking about the `sleep_until` line, it was part of the original asynchronous prognoser example. I just tested with and without the line and noticed no difference, so it's probably safe to remove. It was something I simply glossed over.

Reply:
I'm not sure what's the right answer is here, but I suspect it's trying to accomplish something similar to what `ros::Rate::sleep` would do. If that is so, it'd be best to use the ROS function.

Reply:
I think it was needed in the original example where only one prognoser was declared and the loop did not contain the `waitAll` command (which only ran after the loop terminated). However, since I'm using multiple prognosers and buses here and the `waitAll()` command is within the overall loop, this sleep function isn't actually needed anymore. I could be wrong, but given I saw no noticeable behavior change upon removing the line, that's my hunch.